src/ws/cockpitcertificate.c

```diff
@@ -382,19 +382,32 @@ cockpit_certificate_load (const gchar *cert_path,
   int r;
   g_autofree gchar *certs = NULL;
   g_autofree gchar *key = NULL;
-  g_autofree gchar *combined = NULL;
   GTlsCertificate *cert;
 
-  r = cockpit_certificate_parse (cert_path, &certs, &key);
-  if (r < 0)
+  key = cockpit_certificate_key_path (cert_path);
+  if (key)
     {
-      g_set_error (error, G_IO_ERROR, g_io_error_from_errno (-r), "Failed to load %s: %s", cert_path, g_strerror (-r));
-      return NULL;
+      /* we have a separate .key file */
+      cert = g_tls_certificate_new_from_files (cert_path, key, error);
+      g_debug ("loading separate cert %s and key %s", cert_path, key);
+    }
+  else
+    {
+      /* combined cert+key file */
+      g_autofree gchar *combined = NULL;
+
+      r = cockpit_certificate_parse (cert_path, &certs, &key);
+      if (r < 0)
+        {
+          g_set_error (error, G_IO_ERROR, g_io_error_from_errno (-r), "Failed to load %s: %s", cert_path, g_strerror (-r));
+          return NULL;
+        }
+
+      /* Gio only has constructors for parsing certs and key from one string, so combine them */
```

Comment:
This is weird.  We split the file out and concatenate it again?

We should probably ask for new GIO API here.  Its weird use of the combined format is probably what inspired cockpit's original use of the format and got us into the current mess....

Reply:
The split+recombine is existing code, this PR doesn't touch that case. Admittedly it's rather weird, yes.

`g_tls_certificate_new_from_files()` indeed makes a lot more sense, for the "separate files" case that this PR adds.

Reply:
Speculating: Perhaps `g_tls_certificate_new_from_pem()` expects cert and key in a certain order, and this dance is done to ensure that the key is last? But at least the [current documentation](https://developer.gnome.org/gio/unstable/GTlsCertificate.html#g-tls-certificate-new-from-pem) does not prescribe an ordering.